src/client/workflow/workflow.ts

```diff
@@ -0,0 +1,354 @@
+import { nanoid } from "nanoid";
+import type { AsyncStepFunction, ParallelCallState, Step } from "./types";
+import { WORKFLOW_INTERNAL_HEADER, WORKFLOW_ID_HEADER } from "./constants";
+import type { Client } from "../client";
+import * as WorkflowParser from "./workflow-parser";
+import { AutoExecutor } from "./auto-executor";
+import { QstashWorkflowError } from "../error";
+
+export class Workflow<TInitialRequest = unknown> {
+  protected readonly client: Client;
+  protected executor: AutoExecutor;
+  protected pendingSteps: Step[] = [];
+
+  protected readonly url: string;
+  protected readonly steps: Step[];
+  protected readonly workflowId: string;
+  protected readonly nonPlanStepCount: number;
+  protected skip;
+
+  // counters which are incremented as steps are processed
+  protected stepCount = 0;
+  protected planStepCount = 0;
+
+  public requestPayload: TInitialRequest;
+
+  /**
+   * Creates a workflow context which offers methods to run steps
+   * in parallel and by themselves
+   *
+   * @param client QStash client
+   * @param url QStash backend url
+   * @param workflowId Id of the workflow
+   * @param steps steps received from QStash
+   * @param skip whether the steps in the workflow should be skipped
+   */
+  constructor({
+    client,
+    url,
+    workflowId,
+    steps,
+    skip = false,
+  }: {
+    client: Client;
+    url: string;
+    workflowId: string;
+    steps: Step[];
+    skip: boolean;
+  }) {
+    this.client = client;
+    this.url = url;
+    this.workflowId = workflowId;
+    this.steps = steps;
+    this.nonPlanStepCount = this.steps.filter((step) => !step.targetStep).length;
+    this.skip = skip;
+    this.executor = new AutoExecutor(this);
+    this.requestPayload = this.steps[0].out as TInitialRequest;
+  }
+
+  /**
+   * Executes a workflow step
+   *
+   * ```typescript
+   * const result = context.run("step 1", async () => {
+   *   return await Promise.resolve("result")
+   * })
+   * ```
+   *
+   * Can also be called in parallel and the steps will be executed
+   * simulatenously:
+   *
+   * ```typescript
+   * const [result1, result2] = await Promise.all([
+   *   context.run("step 1", async () => {
+   *     return await Promise.resolve("result1")
+   *   })
+   *   context.run("step 2", async () => {
+   *     return await Promise.resolve("result2")
+   *   })
+   * ])
+   * ```
+   *
+   * - Increments `this.stepCount` by 1.
```

Comment:
These are public documents of the methods that our users will use, right ? I think we should not have `internal` details here. It will confuse the user. These documents can be inside as the first line of the function.

Reply:
Currently this class is what I pass the user but I am planning to change that. There are some public methods I don't want the user to see.

I planning to have a Context class which will have sleepFor, sleepUntil, run methods and requestPayload parameter